modules/local/atacreads/shiftreads.nf

```diff
@@ -0,0 +1,44 @@
+// Import generic module functions
+include { initOptions; saveFiles; getSoftwareName; getProcessName } from '../functions'
+
+params.options = [:]
+options        = initOptions(params.options)
+
+process SHIFTREADS {
+    tag "$meta.id"
+    label 'process_low'
+    publishDir "${params.outdir}",
+        mode: params.publish_dir_mode,
+        saveAs: { filename -> saveFiles(filename:filename, options:params.options, publish_dir:getSoftwareName(task.process), meta:meta, publish_by_meta:['id']) }
+
+    conda (params.enable_conda ? "anaconda::gawk=5.1.0" : null)
+    if (workflow.containerEngine == 'singularity' && !params.singularity_pull_docker_container) {
+        container "https://depot.galaxyproject.org/singularity/gawk:5.1.0"
+    } else {
+        container "quay.io/biocontainers/gawk:5.1.0"
+    }
+
+    input:
+    tuple val(meta), path(pair)
+
+    output:
+    tuple val(meta), path("*.bed.gz"), emit: bed
+    path "versions.yml"           , emit: versions
+
+    script:
+    def software = "awk"
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+    """
+    cat $pair | \\
+        zcat | \\
+        awk 'BEGIN {OFS="\t"};  /^[^#]/ { if (\$7 == "+") {\$5 = \$5 + 4} else if (\$7 == "-") {\$5 = \$5 - 5};  print \$4, \$5, \$5+1, "*", "0", \$7}' | \\
+        sort -k1,1 -k2,2n | \\
+        uniq | \\
+        gzip -nc > ${prefix}.R2.ATAC.bed.gz
```

Comment:
why I always keep using `cat input | zcat` is because it will not work if I use `zcat input` on mac os. In mac OS, it will always add `.Z` to the end of the filename.

Reply:
Understood, but I think the default assumption is that the workflows are intended to be reproducible. As such they should always use a container technology or conda. 

@ewels What is the base assumption of software packaging?

Reply:
Actually, I just realized, that would mean the behaviour is inconsistent. If using `cat` only bypasses the error message, `zcat` should still be performing the same operation regardless if it reads from a file or from a pipe.

Reply:
I will try to use different container to fix this issue.

Reply:
Sorry, I just want be clear here. We should not use `zcat` from MacOS at all here since it's working on `.Z` files. Even though `cat <file> | zcat | ` does not report an error, it means it's still decompressing `.Z` from stdin instead. It is better then, that we use `zcat <file>` so an error is reported if the wrong `zcat` is used.

Reply:
If you use docker/singularity, it is true that the pipeline will use the function in container. It will be different for `conda`. `conda` simply export the PATH. And most of the container for `conda` will not include `zcat`. In MacOS, `zcat` working on `.Z` files. By pipe `cat <file> | zcat` will avoid the filename issue. And the `filename` should not contain wildcard. The performance may be a little different but the output will be same. Correct me if I am not correct.

Reply:
The problem is not a filename issue. The problem is that MacOS `zcat` is trying to uncompress Zip archive data format, which is not the same as gzip. Piping to `zcat` on MacOS then only hides the problem, but still uses Zip archive data decompression instead of gzip decompression.

Reply:
Maybe a better solution is to use `gunzip -c` instead of `zcat`.

Reply:
I will use `gunzip -c` instead. Thank you.